File: Spring/SignUpViewController.m
//
//  SignUpViewController.m
//  Spring
//
//  Created by Joseph Asperger on 10/9/14.
//
//

#import "SignUpViewController.h"
#import "CurrentUserRelations.h"

@interface SignUpViewController ()
@property (weak, nonatomic) IBOutlet UITextField *emailField;
@property (weak, nonatomic) IBOutlet UITextField *usernameField;
@property (weak, nonatomic) IBOutlet UITextField *passwordField;
@property (weak, nonatomic) IBOutlet UITextField *confirmPasswordField;
@property (weak, nonatomic) IBOutlet UILabel *errorLabel;

@end

@implementation SignUpViewController


- (IBAction)signup:(id)sender {
    [self signupUser];
}

- (IBAction)cancel:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

-(void) signupUser{
    if ([self fieldsValid]){
        PFUser *user = [PFUser user];
        user.username = self.emailField.text;
        user.password = self.passwordField.text;
        user[@"displayName"] = self.usernameField.text;
        [user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError *error) {
            if (!error) {
                [CurrentUserRelations downloadRelations];
                [self performSegueWithIdentifier:@"SignupSegue" sender:self];
            } else {
                NSString *errorString = [error userInfo][@"error"];
                self.errorLabel.text = errorString;
            }
        }];
    }
}

-(BOOL) fieldsValid{
    if (self.emailField.text.length == 0){
        self.errorLabel.text = @"Email address missing";
        return NO;
    }
    if (self.usernameField.text.length < 4 || self.usernameField.text.length > 20){
        self.errorLabel.text = @"Username must be 4 to 20 characters";
        return NO;
    }
    if (self.passwordField.text.length < 8 || self.passwordField.text.length > 20){
        self.errorLabel.text = @"Password must be 8 to 20 characters";
        return NO;
    }
    if (![self.passwordField.text isEqualToString:self.confirmPasswordField.text]){
        self.errorLabel.text = @"Passwords do not match";
        return NO;
    }
    return YES;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (textField == self.emailField) {
        [self.usernameField becomeFirstResponder];
    }
    else if (textField == self.usernameField){
        [self.passwordField becomeFirstResponder];
    }
    else if (textField == self.passwordField) {
        [self.confirmPasswordField becomeFirstResponder];
    }
    else if (textField == self.confirmPasswordField) {
        [self.confirmPasswordField resignFirstResponder];
        [self signupUser];
    }
    return YES;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.errorLabel.text = @"";
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void) touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event{
    [self.emailField resignFirstResponder];
    [self.passwordField resignFirstResponder];
    [self.usernameField resignFirstResponder];
    [self.confirmPasswordField resignFirstResponder];
}


/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: Spring/SignUpViewController.swift
//
//  SignUpViewController.swift
//  Spring
//
//  Created by Joseph Asperger on 1/5/15.
//
//

import UIKit

class SignUpViewController: UIViewController, UITextFieldDelegate {
   
    @IBOutlet weak var emailField: UITextField!
    @IBOutlet weak var usernameField: UITextField!
    @IBOutlet weak var passwordField: UITextField!
    @IBOutlet weak var confirmPasswordField: UITextField!
    @IBOutlet weak var errorLabel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.errorLabel.text = ""
    }
    
    @IBAction func signup(sender: AnyObject) {
        signupUser()
    }

    @IBAction func cancel(sender: AnyObject) {
        dismissViewControllerAnimated(true, completion: nil)
    }
    
    func signupUser() {
        if (fieldsValid()){
            var user = PFUser()
            user.username = emailField.text
            user.password = passwordField.text
            user["displayName"] = usernameField.text
            user.signUpInBackgroundWithBlock({ (succeeded, error) -> Void in
                if (succeeded) {
                    CurrentUserRelations.downloadRelations()
                    User.loadEvents()
                self.performSegueWithIdentifier("SignupSegue", sender: self)
                }
            })
        }
    }
    
    func fieldsValid() -> Bool {
        if (countElements(emailField.text) == 0){
            errorLabel.text = "Email address missing"
            return false
        }
        if (countElements(usernameField.text) < 4 || countElements(usernameField.text) > 20){
            errorLabel.text = "Username must be 4 to 20 characters"
            return false
        }
        if (countElements(passwordField.text) < 8 || countElements(passwordField.text) > 20){
            errorLabel.text = "Password must be 8 to 20 characters"
            return false
        }
        if (passwordField.text != confirmPasswordField.text){
            errorLabel.text = "Passwords do not match"
            return false
        }
        return true
    }
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        //move to next text field when user taps next
        if (textField == emailField){
            usernameField.becomeFirstResponder()
        }
        if (textField == usernameField){
            passwordField.becomeFirstResponder()
        }
        if (textField == passwordField){
            confirmPasswordField.becomeFirstResponder()
        }
        if (textField == confirmPasswordField){
            confirmPasswordField.resignFirstResponder()
            signupUser()
        }
        return true
    }
    
    override func touchesBegan(touches: NSSet, withEvent event: UIEvent) {
        // hide keyboard if user taps outside of text fields
        emailField.resignFirstResponder()
        usernameField.resignFirstResponder()
        passwordField.resignFirstResponder()
        confirmPasswordField.resignFirstResponder()
    }
}
